assets/git-ignore.1
'\" t
.\"     Title: git-ignore
.\"    Author: Sondre Nilsen
.\" Generator: Asciidoctor 2.0.16
.\"      Date: 2021-11-21
.\"    Manual: \ \&
.\"    Source: git-ignore v1.2.0
.\"  Language: English
.\"
.TH "GIT\-IGNORE" "1" "2021-11-21" "git\-ignore v1.2.0" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
git-ignore \- quickly and easily list and fetch .gitignore templates from www.gitignore.io
.SH "DESCRIPTION"
.sp
git\-ignore is a small utility to quickly create or add templates from www.gitignore.io
to your .gitignore files. The main difference to other similar utilities is that
it works offline, it does this by caching all available templates and storing them in your
$HOME/.cache directory. It is also reliable and efficient, thanks to Rust.
.SH "SYNOPSIS"
.sp
\fBgit\-ignore\fP [\-V|\-\-version] [\-h|\-\-help] [\-a|\-\-auto] [\-l|\-\-list] [\-s|\-\-simple] [\-u|\-\-update] [subcommand] \fB<templates>\fP
.SH "FLAGS"
.sp
\fB\-h, \-\-help\fP
.RS 4
Print help information
.RE
.sp
\fB\-V, \-\-version\fP
.RS 4
Print version information
.RE
.sp
\fB\-a, \-\-auto\fP
.RS 4
Automatically infer templates based on files in the current directory.
.RE
.sp
\fB\-l, \-\-list\fP
.RS 4
List your selected templates, or all available
.RE
.sp
\fB\-s, \-\-simple\fP
.RS 4
Ignore your user defined aliases and templates
.RE
.sp
\fB\-u, \-\-update\fP
.RS 4
Update the local cache from www.gitignore.io
.RE
.SH "COMMANDS"
.sp
\fBgit\-ignore init\fP
.RS 4
Initialize user configuration
.RE
.sp
\fBgit\-ignore completion\fP
.RS 4
Generate shell completion
.RE
.sp
\fBgit\-ignore alias list\fP
.RS 4
List your created aliases
.RE
.sp
\fBgit\-ignore alias add <name> <aliases>\fP
.RS 4
Add a new alias that maps to a list of one or more other templates
.RE
.sp
\fBgit\-ignore alias remove <name>\fP
.RS 4
Remove an alias
.RE
.sp
\fBgit\-ignore template list\fP
.RS 4
List your created templates
.RE
.sp
\fBgit\-ignore template add <name> <path>\fP
.RS 4
Add a new template, note that the path is just the file name and should be \fBhaskell.txt\fP and not
a full path.
.RE
.sp
\fBgit\-ignore template remove <name>\fP
.RS 4
Remove a template.
.RE
.SH "USAGE NOTES"
.sp
If the required \f(CRignore.json\fP file does not exist, an attempt will be made to
download it. This requires an internet connection, but only once.
.sp
The program will print a small notice when you are only using cached templates,
this is printed to STDERR so it will not interfere with piping etc.",
.sp
Note that listing templates doesn\(cqt require exact matches while printing the
template does. When listing it matches any template starting with each query,
i.e. \f(CRintellij\fP matches all templates starting with \f(CRintellij\fP (see example below).
.SH "EXAMPLES"
.sp
List all available templates
.RS 4
$ git ignore \-\-list
.br
[ list of all templates ]
.RE
.sp
List all matching \fB<templates>\fP
.RS 4
$ git ignore \-\-list rust intellij
.br
[ rust intellij intellij+all intellij+iml ]
.RE
.sp
Print matching templates to STDOUT
.RS 4
$ git ignore rust intellij
.br
# RUST # [...]
.RE
.sp
Add an alias for \fBrust\fP to \fBrust, visualstudiocode\fP
.RS 4
$ git ignore alias add rust rust visualstudiocode
.br
Created alias rust for ["rust", "visualstudiocode"]
.RE
.SH "VERSION"
.sp
v1.2.0
.SH "RESOURCES"
.sp
\fBProject web site:\fP \c
.URL "https://github.com/sondr3/git\-ignore" "" ""
.sp
\fBSource code:\fP \c
.URL "https://github.com/sondr3/git\-ignore" "" ""
.SH "COPYING"
.sp
Free use of this software is granted under the terms of either the MIT or Apache\-2.0 license.
.SH "AUTHOR"
.sp
Sondre Nilsen


CHANGELOG.md
## 1.2.0

> 2021-12-05

The biggest release yet! From automatically recognizing templates based on files
in your current directory to custom aliases and templates! Huge thanks to 
[`sestrella`](https://github.com/sestrella) for contributing the automatic matching functionality.

- **Automatic**: When using the `-a/--auto`, `git-ignore` will automatically create `.gitignore`
  templates by matching against files in your current directory.
- **Aliases**: Do you use the same templates toghether all the time? Create an alias to combine
  them.
- **Templates**: Missing a template from www.gitignore.io? Create your own!

### Commits

* [[`75a40daa25`](https://github.com/sondr3/git-ignore/commit/75a40daa25)] - Upgrade to clap v3 RC 
* [[`1d9c331479`](https://github.com/sondr3/git-ignore/commit/1d9c331479)] - Add a bunch more languages to detectors 
* [[`688b672d77`](https://github.com/sondr3/git-ignore/commit/688b672d77)] - \[#7] Auto detect templates based on existing files
* [[`33af5b4022`](https://github.com/sondr3/git-ignore/commit/33af5b4022)] - Add notice to config\_or! macro when config is missing 
* [[`cadc69d516`](https://github.com/sondr3/git-ignore/commit/cadc69d516)] - Add visible aliases to subcommands, update man page 
* [[`738deff37d`](https://github.com/sondr3/git-ignore/commit/738deff37d)] - Early exit if updating and not listing any templates 
* [[`7a56a7ebc8`](https://github.com/sondr3/git-ignore/commit/7a56a7ebc8)] - Add completions to assets folder 
* [[`1eefbac6b8`](https://github.com/sondr3/git-ignore/commit/1eefbac6b8)] - Remove man crate, create man pages via asciidoctor 
* [[`f8e3c63453`](https://github.com/sondr3/git-ignore/commit/f8e3c63453)] - Update man page, fix some spelling 
* [[`a3c1124c48`](https://github.com/sondr3/git-ignore/commit/a3c1124c48)] - Minor error handling refactoring 
* [[`72a72733f9`](https://github.com/sondr3/git-ignore/commit/72a72733f9)] - Add anyhow for easier error handling 
* [[`e97aca5ab9`](https://github.com/sondr3/git-ignore/commit/e97aca5ab9)] - Add --simple flag to ignore aliases and templates 
* [[`dc748b2cbc`](https://github.com/sondr3/git-ignore/commit/dc748b2cbc)] - Implement creating and reading templates 
* [[`8f767e9078`](https://github.com/sondr3/git-ignore/commit/8f767e9078)] - Simplify and remove unneeded code 
* [[`dc7dacd018`](https://github.com/sondr3/git-ignore/commit/dc7dacd018)] - Color aliases, templates when listing 
* [[`660e99085a`](https://github.com/sondr3/git-ignore/commit/660e99085a)] - Fix some pedantic lints 
* [[`3b7cbc5072`](https://github.com/sondr3/git-ignore/commit/3b7cbc5072)] - Switch lookup of names from starts\_with to contains 
* [[`c9b49eb02d`](https://github.com/sondr3/git-ignore/commit/c9b49eb02d)] - Fetch templates from aliases and list them/print them 
* [[`9ddfa8d2df`](https://github.com/sondr3/git-ignore/commit/9ddfa8d2df)] - Refactor Config::from\_dir function 
* [[`d558dd1ebc`](https://github.com/sondr3/git-ignore/commit/d558dd1ebc)] - Fix clippy warning 
* [[`2f8cfebcb1`](https://github.com/sondr3/git-ignore/commit/2f8cfebcb1)] - Add messages when removing aliases and templates 
* [[`9b3b74714e`](https://github.com/sondr3/git-ignore/commit/9b3b74714e)] - Fix error where config does not get its own path 
* [[`9a6925357a`](https://github.com/sondr3/git-ignore/commit/9a6925357a)] - Simplify config initialization, don't list empty aliases 
* [[`9a91755aae`](https://github.com/sondr3/git-ignore/commit/9a91755aae)] - Extract handling of config to a macro 
* [[`5719bfdf05`](https://github.com/sondr3/git-ignore/commit/5719bfdf05)] - Add methods for listing templates and aliases 
* [[`71c7f1c5a7`](https://github.com/sondr3/git-ignore/commit/71c7f1c5a7)] - Extract GitIgnore struct to own file 
* [[`e5f32a119e`](https://github.com/sondr3/git-ignore/commit/e5f32a119e)] - Extract Config to own file, add methods for adding/removing aliases 
* [[`6751e35588`](https://github.com/sondr3/git-ignore/commit/6751e35588)] - Use Clap from GitHub while waiting for next release 
* [[`ff2248b602`](https://github.com/sondr3/git-ignore/commit/ff2248b602)] - Extract CLI, add completions 
* [[`f761896936`](https://github.com/sondr3/git-ignore/commit/f761896936)] - Initial port of commands from structopt to clap v3 
* [[`5b2a3b8077`](https://github.com/sondr3/git-ignore/commit/5b2a3b8077)] - Simplify getting config directory 
* [[`4c8746961e`](https://github.com/sondr3/git-ignore/commit/4c8746961e)] - Add handling for templates 
* [[`73f5cb59b9`](https://github.com/sondr3/git-ignore/commit/73f5cb59b9)] - Convert config handling to subcommands 
* [[`6ef7eaa0ce`](https://github.com/sondr3/git-ignore/commit/6ef7eaa0ce)] - Implement adding aliases 
* [[`d5b6793b69`](https://github.com/sondr3/git-ignore/commit/d5b6793b69)] - Start working on a config file 
* [[`f966329d98`](https://github.com/sondr3/git-ignore/commit/f966329d98)] - Upgrade to Rust 2021 edition, upgrade packages 

## 1.1.2

> 2021-03-24

No functionality added this release, upgraded dependencies and fixed manpage generation.

- [[`7573fe6c2a`](https://github.com/sondr3/git-ignore/commit/7573fe6c2acbeaafd5c4762d63ead04cfe8de523)] Release 1.1.2
- [[`2d4fe00657`](https://github.com/sondr3/git-ignore/commit/2d4fe00657d144f3be3ffff563c671910fda7e0a)] Update all dependencies, format Cargo.toml
- [[`51afde98be`](https://github.com/sondr3/git-ignore/commit/51afde98be7eb0c0712514a21def1a6312ad38b2)] Remove test CI pipeline, there are no tests in this project
- [[`18e5428950`](https://github.com/sondr3/git-ignore/commit/18e5428950b78cfcd984da4a8d2dc580bbf97fae)] Fix lint errors
- [[`a51bfdeb62`](https://github.com/sondr3/git-ignore/commit/a51bfdeb627943785dd062dbd00168ba0d9ccf6f)] Copy and add man page to release archives
- [[`b889b975da`](https://github.com/sondr3/git-ignore/commit/b889b975da3e10a57ef9bdd874093f1212780057)] Upgrade dependencies

assets/git-ignore.bash
_git-ignore() {
    local i cur prev opts cmds
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${i}" in
            "$1")
                cmd="git__ignore"
                ;;
            add)
                cmd+="__add"
                ;;
            alias)
                cmd+="__alias"
                ;;
            completion)
                cmd+="__completion"
                ;;
            help)
                cmd+="__help"
                ;;
            init)
                cmd+="__init"
                ;;
            list)
                cmd+="__list"
                ;;
            remove)
                cmd+="__remove"
                ;;
            template)
                cmd+="__template"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        git__ignore)
            opts="-h -V -l -u -s -a --help --version --list --update --simple --auto <TEMPLATES>... alias template init completion help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__alias)
            opts="-h --help list add remove help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__alias__add)
            opts="-h --version --help <NAME> <ALIASES>..."
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__alias__help)
            opts="-h --version --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__alias__list)
            opts="-h --version --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__alias__remove)
            opts="-h --version --help <NAME>"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__completion)
            opts="-h --help bash elvish fish powershell zsh"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__help)
            opts="-h --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__init)
            opts="-h --force --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__template)
            opts="-h --help list add remove help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__template__add)
            opts="-h --version --help <NAME> <FILE_NAME>"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__template__help)
            opts="-h --version --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__template__list)
            opts="-h --version --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        git__ignore__template__remove)
            opts="-h --version --help <NAME>"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

complete -F _git-ignore -o bashdefault -o default git-ignore
